internal/healthchecks/builtin.go

```diff
@@ -0,0 +1,170 @@
+package healthchecks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AreObjectsHealthy checks if the given resources are healthy.
+// It returns a nil error if all the resources are healthy, if any resource is not healthy, the error will
+// contain the GVK + namespace/resourceName and the error message of each unhealthy resource.
+//
+// The current list of supported resources is:
+// - Deployments
+// - StatefulSets
+// - DaemonSets
+// - ReplicaSets
+// - Pods
+// - APIServices
+// - CustomResourceDefinitions
+//
+// If the resource is not supported, it is assumed to be healthy.
+func AreObjectsHealthy(ctx context.Context, client client.Client, objects []client.Object) error {
+	var gvkErrors []error
+
+	for _, object := range objects {
+		// Convert the object to its proper type, deepCopy it to avoid mutating the original object.
+		obj, err := runtime.ObjectConvertor(client.Scheme()).ConvertToVersion(object.DeepCopyObject(), object.GetObjectKind().GroupVersionKind().GroupVersion())
+		if err != nil {
+			gvkErrors = appendResourceError(gvkErrors, object, err.Error())
+			continue
+		}
+
+		objectKey := types.NamespacedName{
+			Name:      object.GetName(),
+			Namespace: object.GetNamespace(),
+		}
+
+		switch obj := obj.(type) {
+		case *appsv1.Deployment:
```

Comment:
Global comment: Helm's similar case statements include multiple versions of the same GroupKind. If I'm understanding this correctly, if a bundle has an `apps/v1beta1` Deployment in it, this code will always report back "healthy" for that object.

https://github.com/helm/helm/blob/e7bb860d9a32e8739c944b8e7b7f7031d752411a/pkg/kube/ready.go#L115

Reply:
yes, we are checking for the v1 version only, I can add logic for the `apps/v1beta1`. Any other you can think of?

Reply:
@ncdc I need help finding a way to satisfy the interface for client.Get(), can you help me understand how unstructured can be used here?

Reply:
Try this
```
u := &unstructured.Unstructured{}
		u.SetGroupVersionKind(object.GetObjectKind().GroupVersionKind())
		if err := client.Get(ctx, objectKey, u); err != nil {
			
		}
```

Reply:
Thanks @ncdc, PTAL